### ISSUE TITLE
`Internal Server Error` when disabling `Auto-Forward` in the admin console

### ISSUE THREAD

**Issue by tomasz-zieja** [Mailu/Mailu#3352]
## Environment & Version
- Docker Compose version `v2.27.1`
- Version: `2024.06`

## Description
`Internal Server Error` when disabling `Auto-Forward` in the admin console.

## Replication Steps
1) Log in as a user to the admin console.
2) Go to `Settings`.
3) Tick `Enable Forwarding` box, type a forward destination address.
4) Save settings.
5) Try to disable `Auto-Forward` by unticking the `Enable Forwarding` box.
6) Save settings.

## Observed behaviour
`Internal Server Error` is displayed in a web browser.

## Expected behaviour
`Auto-Forward` should be disabled.

## Workaround
Type your own email address in the `Destination` text field and keep the `Enable Forwarding` box ticked.

## Logs
```
lip 27 13:31:10 mailu-admin[525]: [2024-07-27 13:31:10,017] ERROR in app: Exception on /admin/user/settings [POST]
lip 27 13:31:10 mailu-admin[525]: Traceback (most recent call last):
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1815, in _execute_context
lip 27 13:31:10 mailu-admin[525]:     context = constructor(
lip 27 13:31:10 mailu-admin[525]:               ^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 1459, in _init_compiled
lip 27 13:31:10 mailu-admin[525]:     flattened_processors[key](compiled_params[key])
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/sql/type_api.py", line 2044, in process
lip 27 13:31:10 mailu-admin[525]:     return fixed_process_param(value, dialect)
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/mailu/models.py", line 90, in process_bind_param
lip 27 13:31:10 mailu-admin[525]:     raise TypeError('Must be a list of strings')
lip 27 13:31:10 mailu-admin[525]: TypeError: Must be a list of strings
lip 27 13:31:10 mailu-admin[525]:
lip 27 13:31:10 mailu-admin[525]: The above exception was the direct cause of the following exception:
lip 27 13:31:10 mailu-admin[525]:
lip 27 13:31:10 mailu-admin[525]: Traceback (most recent call last):
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
lip 27 13:31:10 mailu-admin[525]:     response = self.full_dispatch_request()
lip 27 13:31:10 mailu-admin[525]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
lip 27 13:31:10 mailu-admin[525]:     rv = self.handle_user_exception(e)
lip 27 13:31:10 mailu-admin[525]:          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
lip 27 13:31:10 mailu-admin[525]:     rv = self.dispatch_request()
lip 27 13:31:10 mailu-admin[525]:          ^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
lip 27 13:31:10 mailu-admin[525]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/flask_login/utils.py", line 290, in decorated_view
lip 27 13:31:10 mailu-admin[525]:     return current_app.ensure_sync(func)(*args, **kwargs)
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/mailu/ui/access.py", line 28, in wrapper
lip 27 13:31:10 mailu-admin[525]:     return callback(function, args, kwargs, dargs, dkwargs)
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/mailu/ui/access.py", line 19, in callback
lip 27 13:31:10 mailu-admin[525]:     return function(*args, **kwargs)
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/mailu/ui/views/users.py", line 103, in user_settings
lip 27 13:31:10 mailu-admin[525]:     models.db.session.commit()
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/scoping.py", line 597, in commit
lip 27 13:31:10 mailu-admin[525]:     return self._proxied.commit()
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2017, in commit
lip 27 13:31:10 mailu-admin[525]:     trans.commit(_to_root=True)
lip 27 13:31:10 mailu-admin[525]:   File "<string>", line 2, in commit
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
lip 27 13:31:10 mailu-admin[525]:     ret_value = fn(self, *arg, **kw)
lip 27 13:31:10 mailu-admin[525]:                 ^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 1302, in commit
lip 27 13:31:10 mailu-admin[525]:     self._prepare_impl()
lip 27 13:31:10 mailu-admin[525]:   File "<string>", line 2, in _prepare_impl
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
lip 27 13:31:10 mailu-admin[525]:     ret_value = fn(self, *arg, **kw)
lip 27 13:31:10 mailu-admin[525]:                 ^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 1277, in _prepare_impl
lip 27 13:31:10 mailu-admin[525]:     self.session.flush()
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4341, in flush
lip 27 13:31:10 mailu-admin[525]:     self._flush(objects)
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4476, in _flush
lip 27 13:31:10 mailu-admin[525]:     with util.safe_reraise():
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
lip 27 13:31:10 mailu-admin[525]:     raise exc_value.with_traceback(exc_tb)
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4437, in _flush
lip 27 13:31:10 mailu-admin[525]:     flush_context.execute()
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
lip 27 13:31:10 mailu-admin[525]:     rec.execute(self)
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
lip 27 13:31:10 mailu-admin[525]:     util.preloaded.orm_persistence.save_obj(
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 85, in save_obj
lip 27 13:31:10 mailu-admin[525]:     _emit_update_statements(
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 912, in _emit_update_statements
lip 27 13:31:10 mailu-admin[525]:     c = connection.execute(
lip 27 13:31:10 mailu-admin[525]:         ^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
lip 27 13:31:10 mailu-admin[525]:     return meth(
lip 27 13:31:10 mailu-admin[525]:            ^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
lip 27 13:31:10 mailu-admin[525]:     return connection._execute_clauseelement(
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
lip 27 13:31:10 mailu-admin[525]:     ret = self._execute_context(
lip 27 13:31:10 mailu-admin[525]:           ^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1821, in _execute_context
lip 27 13:31:10 mailu-admin[525]:     self._handle_dbapi_exception(
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
lip 27 13:31:10 mailu-admin[525]:     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1815, in _execute_context
lip 27 13:31:10 mailu-admin[525]:     context = constructor(
lip 27 13:31:10 mailu-admin[525]:               ^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 1459, in _init_compiled
lip 27 13:31:10 mailu-admin[525]:     flattened_processors[key](compiled_params[key])
lip 27 13:31:10 mailu-admin[525]:   File "/app/venv/lib/python3.12/site-packages/sqlalchemy/sql/type_api.py", line 2044, in process
lip 27 13:31:10 mailu-admin[525]:     return fixed_process_param(value, dialect)
lip 27 13:31:10 mailu-admin[525]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lip 27 13:31:10 mailu-admin[525]:   File "/app/mailu/models.py", line 90, in process_bind_param
lip 27 13:31:10 mailu-admin[525]:     raise TypeError('Must be a list of strings')
lip 27 13:31:10 mailu-admin[525]: sqlalchemy.exc.StatementError: (builtins.TypeError) Must be a list of strings
lip 27 13:31:10 mailu-admin[525]: [SQL: UPDATE user SET forward_enabled=?, forward_destination=?, forward_keep=?, updated_at=?, email=? WHERE user.email = ?]
lip 27 13:31:10 mailu-admin[525]: [parameters: [{'forward_destination': 'forward@domain.com', 'forward_keep': False, 'forward_enabled': False, 'user_email': 'example@domain.com'}]]
```

**Comment by toxanes** [comment]
I have the same error, I can't fix it. I also managed to repeat it on test.mailu.io

**Comment by Wombosvideo** [comment]
Duplicate of #3349

**Comment by nextgens** [comment]
Please try the following and report back if it fixes it:

```
export MAILU_VERSION=pr-3353
docker compose pull
docker compose up -d
```

**Comment by tomasz-zieja** [comment]
With the PR it is impossible to enable `Auto-Forward` and the interface ends up in an inconsistent state.

Steps to reproduce:
1) Log in as a user to the admin console.
2) Go to `Settings`.
3) Tick the `Enable Forwarding` box, type a forward destination address.
4) Click `Save settings`.
5) Go to `Settings` again to refresh the page.

Notice that the interface ends up in an inconsistent state:
- The `Enable forwarding` box is not ticked but...
- The `Keep a copy of the emails` box is not grayed out.
- The `Destination` field is not greyed out.

Furthermore, the `Auto-Forward` feature doesn't work, emails are not forwarded.

**Comment by tomasz-zieja** [comment]
![Auto-Forward](https://github.com/user-attachments/assets/f65a9b5d-2d3a-4ea2-a379-4e183f6997ff)

**Comment by tomasz-zieja** [comment]
I made a test and it seems that auto forwarding does work with the PR. So the problem might be just with the interface.

**Comment by tomasz-zieja** [comment]
The behavior is confusing. When you enable auto forwarding with `pr-3353`, then it doesn't forward emails to the destination address in addition to the interface being broken. But after you simply go back to `2024.06`, even without making any changes to the settings, then it suddenly starts forwarding them again. Something strange is going on there. I hope it's helpful.

**Comment by nextgens** [comment]
@tomasz-zieja thank you for trying it out

Please re-pull the image and retry. I have added a new commit that ought to fix it

**Comment by tomasz-zieja** [comment]
Looks like everything works fine now! Thanks. Are you going to include the fix in `2024.06`?

**Comment by nextgens** [comment]
Thank you for double-checking. Yes, once it's reviewed it will make it to the next point release.
